src/content/docs/en/recipes/making-toolbar-apps.mdx

```diff
@@ -0,0 +1,229 @@
+
+export default defineToolbarApp({
+    init(canvas, app, server) {
+      // Create a new text node every time the app is toggled on or off
+      app.onAppToggled(({ state }) => {
+        const text = document.createTextNode(
+          `The app is now ${state ? "enabled" : "disabled"}!`,
+        );
+        canvas.appendChild(text);
+      });
+      
+      // Send a message to the server
+      server.send("my-app:hello-server", {message: "Hello, server!"});
+    },
+});
+```
+
+These arguments provide multiple utilities and helpers to interact with the dev toolbar and the server, such as executing code when the app is toggled on or off, showing a notification dot on the app, sending messages to the server, and more.
+
+For more information on the hooks and helpers available in the `init()` function, see the [Astro Toolbar API documentation](/en/reference/dev-toolbar-app-reference/).
+
+## Using a UI framework
+
+### Without a build step
+
+If your framework supports it, you can create a dev toolbar app using a framework without a build step. For example, you can use Preact's `h` function to create elements and render them directly to the ShadowRoot.
+
+```ts title="app.ts"
+import { defineToolbarApp } from "astro/toolbar";
+import { render, h } from "preact";
+
+export default defineToolbarApp({
+    init(canvas) {
+        render(h('h1', null, 'Hello World!'), canvas);
+    },
+});
+```
+
+The [HTM package](https://github.com/developit/htm) is a good choice for creating dev toolbar apps without a build step, offering native integration for React and Preact and support for other frameworks.
+
+```ts title="app.ts" ins={3, 7}
+import { defineToolbarApp } from "astro/toolbar";
+import { render } from "preact";
+import { html } from 'htm/preact';
+
+export default defineToolbarApp({
+    init(canvas) {
+        render(html`<h1>Hello World!</h1>`, canvas);
+    },
+});
+```
+
+### With a build step
+
+Astro does not preprocess JSX code in dev toolbar apps. As such, a build step is required in order to use JSX components in your dev toolbar app. In this recipe, we'll be using TypeScript to do this, but any other tools that compile JSX code will work as well (e.g. Babel, Rollup, ESBuild).
+
+First, install TypeScript inside your project:
+
+<PackageManagerTabs>
+  <Fragment slot="npm">
+    ```shell
+    npm install --save-dev typescript
+    ```
+  </Fragment>
+  <Fragment slot="pnpm">
+    ```shell
+    pnpm install --save-dev typescript
+    ```
+  </Fragment>
+  <Fragment slot="yarn">
+    ```shell
+    yarn add --dev typescript
+    ```
+  </Fragment>
+</PackageManagerTabs>
+
+Next, create a `tsconfig.json` file in the root of your project with the appropriate settings to build and for the framework you're using ([React](https://react-typescript-cheatsheet.netlify.app/docs/basic/setup), [Preact](https://preactjs.com/guide/v10/typescript), [Solid](https://www.solidjs.com/guides/typescript)). For example, for Preact:
```

Comment:
I read "root of your project" as top-level. Would "root of your toolbar app" make it more clear?